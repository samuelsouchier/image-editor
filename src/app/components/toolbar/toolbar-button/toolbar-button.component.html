<button
  class="toolbar-tool"
  [tooltip]="tooltipLabel"
  [position]="tooltipPosition"
  [disabled]="disabled"
  (click)="buttonClick.emit()"
  (keyup.enter)="buttonClick.emit()"
>
  <app-icon [iconName]="icon" />
  <span *ngIf="slider" class="toolbar-tool-slider">
    <input type="range" class="toolbar-tool-slider-input" (change)="onChange($event)" [min]="sliderMin" [max]="sliderMax" [value]="sliderValue">
  </span>
</button>
