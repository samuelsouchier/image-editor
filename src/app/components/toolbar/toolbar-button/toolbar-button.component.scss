@import 'src/styles/variables';
@import 'src/styles/colors';

.toolbar-tool {
  position: relative;
  cursor: pointer;
  padding: $spacing-xs;
  border-radius: $border-radius;
  display: inline-block;
  width: fit-content;
  height: fit-content;
  transition: .2s ease-in;
  background: transparent;
  border: none;
  color: $color-primary;
  &:disabled {
    color: $color-disabled;
    pointer-events: none;
  }
  &-slider {
    opacity: 0;
    z-index: -9;
    display: none;
    position: absolute;
    padding: $spacing-xs $spacing-s;
    border-radius: $border-radius;
    background: $color-white;
    border: 1px solid $color-primary;
    bottom: 0;
    left: 50%;
    transform: translate(-50%, 100%);
    transition: .2s ease-in;
    &-input {
      -webkit-appearance: none;
      appearance: none;
      background: transparent;
      cursor: pointer;
      &:focus {
        outline: none;
      }

      &::-webkit-slider-runnable-track {
        background: $color-primary;
        height: $spacing-xxs;
        border-radius: $border-radius;
      }
      &::-moz-range-track {
        background: $color-primary;
        height: $spacing-xxs;
        border-radius: $border-radius;
      }
      &::-webkit-slider-thumb {
        -webkit-appearance: none;
        appearance: none;
        background-color: $color-accent;
        height: $spacing-s;
        width: $spacing-s;
        border-radius: 50%;
        transform: translateY($spacing-xxs * -0.75);
      }
      &::-moz-range-thumb {
        border: none;
        background-color: $color-accent;
        height: $spacing-s;
        width: $spacing-s;
        border-radius: 50%;
        transform: translateY($spacing-xxs * -0.75);
      }
    }
  }
  &:hover {
    background: rgba($color-primary, 0.2);
  }
  &:focus-within {
    .toolbar-tool-slider {
      z-index: 9;
      opacity: 1;
      display: block;
    }
  }
}
