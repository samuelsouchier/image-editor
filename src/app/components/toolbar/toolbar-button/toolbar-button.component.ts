import { CommonModule } from '@angular/common';
import { Component, EventEmitter, Input, Output } from '@angular/core';
import { TooltipDirective } from '../../../directives/tooltip.directive';
import { IconName } from '../../../model/icon-name';
import { TooltipPosition } from '../../../model/tooltip-position';
import { IconComponent } from '../../icon/icon.component';

@Component({
  selector: 'app-toolbar-button',
  standalone: true,
  imports: [CommonModule, IconComponent, TooltipDirective],
  templateUrl: './toolbar-button.component.html',
  styleUrls: ['./toolbar-button.component.scss']
})
export class ToolbarButtonComponent {
  @Input({required: true}) icon: IconName;
  @Input({required: true}) tooltipLabel: string;
  @Input({required: true}) tooltipPosition: TooltipPosition;
  @Input() disabled = false;
  @Input() slider = false;
  @Input() sliderMin = 0;
  @Input() sliderMax = 200;
  @Input() sliderValue = 100;

  @Output() buttonClick = new EventEmitter<void>();
  @Output() sliderChange = new EventEmitter<number>();


  onChange(event: Event) {
    const inputValue = this.#getInputValue(event);
    if (inputValue) {
      this.sliderChange.emit(parseInt(inputValue) / 100);
    }
  }
  #getInputValue(event: Event): string {
    return (event.target as HTMLInputElement).value;
  }
}
