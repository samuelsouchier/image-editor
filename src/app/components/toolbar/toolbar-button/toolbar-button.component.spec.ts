import { createComponentFactory, Spectator } from '@ngneat/spectator/jest';

import { ToolbarButtonComponent } from './toolbar-button.component';

describe(ToolbarButtonComponent.name, () => {
  let spectator: Spectator<ToolbarButtonComponent>;
  const createComponent = createComponentFactory({
    component: ToolbarButtonComponent,
    shallow: true,
  });

  it('should match snapshot without a slider', () => {
    spectator = createComponent(
      {
        props: {
          icon: 'image',
          tooltipLabel: 'tooltip label',
          tooltipPosition: 'top',
        }
      }
    );

    expect(spectator.fixture).toMatchSnapshot();
  });
});
