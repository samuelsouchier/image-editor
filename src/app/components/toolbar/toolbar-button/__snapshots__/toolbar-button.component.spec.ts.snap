// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ToolbarButtonComponent should match snapshot without a slider 1`] = `
<app-toolbar-button
  buttonClick={[Function EventEmitter_]}
  disabled="false"
  icon={[Function String]}
  slider="false"
  sliderChange={[Function EventEmitter_]}
  sliderMax={[Function Number]}
  sliderMin="0"
  sliderValue={[Function Number]}
  tooltipLabel={[Function String]}
  tooltipPosition={[Function String]}
>
  <button
    class="toolbar-tool"
  >
    <app-icon>
      <div
        class="icon"
      >
        <svg
          class="feather feather-image"
          fill="none"
          height="24"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          viewBox="0 0 24 24"
          width="24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <rect
            height="18"
            rx="2"
            ry="2"
            width="18"
            x="3"
            y="3"
          />
          <circle
            cx="8.5"
            cy="8.5"
            r="1.5"
          />
          <polyline
            points="21 15 16 10 5 21"
          />
        </svg>
      </div>
    </app-icon>
  </button>
</app-toolbar-button>
`;
